File: services/api/src/main/java/com/voltpilot/api/uems/GrenzeAufloesung.java
package com.voltpilot.api.uems;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Comparator;
import java.util.List;

/**
 * Die EINE Lese-Regel der Grenzen am Netzanschluss (UEMS AP-15 IP-3, Kasten W1, Konzept §6.1): je Richtung gilt
 * der ENGERE Wert aus Anlage und Netzanschluss; ohne gebundenen Netzanschluss oder ohne gültige Fassung des
 * Grenzblatts der alte Wert der Anlage — dasselbe Objekt, damit eine Anlage ohne Eintrag Byte für Byte bleibt.
 *
 * <p>Rein: ohne Spring, ohne DB, ohne Uhr. Der Python-Zwilling ist
 * {@code services/optimization/voltpilot_optimization/grenze_aufloesung.py}; beide fahren
 * {@code docs/contracts/v2/netzanschluss-grenze-vectors.json}. <b>Wer die Regel ändert, ändert die Vektor-Datei UND
 * beide Zwillinge.</b>
 *
 * <p>Die Werte der Anlage: Einspeisung = {@code site.max_feed_in_kw}, Bezug = {@code site_charging_config.grid_limit_kw}
 * (die Anschlussgrenze des Ladeparks). Tarif und Vergütung ziehen NICHT um (W1).
 */
public final class GrenzeAufloesung {

    public static final String QUELLE_ANLAGE = "anlage";
    public static final String QUELLE_NETZANSCHLUSS = "netzanschluss";

    public static final String FEHLER_UEBER_VEREINBART = "grenze_ueber_vereinbart";
    public static final String FEHLER_UEBER_ANSCHLUSS = "grenze_ueber_anschluss";

    private GrenzeAufloesung() {}

    /** Die zwei Grenzen einer Seite; {@code null} = keine Grenze bekannt (nie 0 kW). */
    public record Grenzen(BigDecimal einspeisungKw, BigDecimal bezugKw) {}

    /** Eine Fassung des Grenzblatts: gilt ab ihrem Tag bis zum Vortag der nächsten. */
    public record Fassung(LocalDate gueltigAb, BigDecimal einspeisegrenzeKw, BigDecimal bezugsgrenzeKw,
            boolean einspeisegrenzeKeine) {
        public Fassung(LocalDate gueltigAb, BigDecimal einspeisegrenzeKw, BigDecimal bezugsgrenzeKw) {
            this(gueltigAb, einspeisegrenzeKw, bezugsgrenzeKw, false);
        }
    }

    /** Wert oder unbekannt; nur Einspeisung darf ausdrücklich unbegrenzt sein (I1). Quelle bleibt dabei das Blatt. */
    public record Wirksam(BigDecimal einspeisungKw, String quelleEinspeisung, BigDecimal bezugKw, String quelleBezug,
            boolean einspeisungKeine) {
        public boolean grenzenGesetzt() {
            return bezugKw != null && (einspeisungKw != null || einspeisungKeine);
        }
    }

    /** Die Fassung, die am {@code tag} gilt: die mit dem spätesten ersten Tag ≤ {@code tag}; sonst {@code null}. */
    public static Fassung fassungAm(List<Fassung> fassungen, LocalDate tag) {
        if (fassungen == null || tag == null) {
            return null;
        }
        return fassungen.stream()
                .filter(f -> !f.gueltigAb().isAfter(tag))
                .max(Comparator.comparing(Fassung::gueltigAb))
                .orElse(null);
    }

    /**
     * Die wirksamen Grenzen am {@code tag}. {@code gebunden} sagt, ob die Anlage an dem Tag an einem Netzanschluss
     * hängt; {@code fassungen} sind die wirksamen Fassungen DIESES Anschlusses.
     */
    public static Wirksam aufloesen(Grenzen anlage, boolean gebunden, List<Fassung> fassungen, LocalDate tag) {
        BigDecimal einspeisung = anlage == null ? null : anlage.einspeisungKw();
        BigDecimal bezug = anlage == null ? null : anlage.bezugKw();
        Fassung f = gebunden ? fassungAm(fassungen, tag) : null;
        BigDecimal naEinspeisung = f == null ? null : f.einspeisegrenzeKw();
        BigDecimal naBezug = f == null ? null : f.bezugsgrenzeKw();
        boolean keine = einspeisung == null && f != null && f.einspeisegrenzeKeine();
        return new Wirksam(
                engerer(einspeisung, naEinspeisung), keine ? QUELLE_NETZANSCHLUSS : quelle(einspeisung, naEinspeisung),
                engerer(bezug, naBezug), quelle(bezug, naBezug), keine);
    }

    /** Der engere Wert; gleich = der Wert der Anlage (dasselbe Objekt); einer fehlt = der andere. */
    public static BigDecimal engerer(BigDecimal anlage, BigDecimal netzanschluss) {
        if (netzanschluss == null) {
            return anlage;
        }
        if (anlage == null) {
            return netzanschluss;
        }
        return netzanschluss.compareTo(anlage) < 0 ? netzanschluss : anlage;
    }

    private static String quelle(BigDecimal anlage, BigDecimal netzanschluss) {
        BigDecimal wert = engerer(anlage, netzanschluss);
        if (wert == null) {
            return null;
        }
        return wert == anlage ? QUELLE_ANLAGE : QUELLE_NETZANSCHLUSS;
    }

    /**
     * Die Plausibilität einer Fassung gegen den Anschluss (AP-01 E10, bekannte Regel des Ladepark-Rahmens): die
     * Bezugsgrenze liegt nie über der vereinbarten Leistung, keine Grenze über der Anschlussleistung (kW ≤ kVA bei
     * cos φ = 1). Fehlt der Vergleichswert, gibt es nichts zu prüfen. {@code null} = plausibel, sonst der Code.
     */
    public static String plausibel(BigDecimal einspeisegrenzeKw, BigDecimal bezugsgrenzeKw, BigDecimal vereinbartKw,
            BigDecimal anschlussKva) {
        if (bezugsgrenzeKw != null && vereinbartKw != null && bezugsgrenzeKw.compareTo(vereinbartKw) > 0) {
            return FEHLER_UEBER_VEREINBART;
        }
        if (anschlussKva != null && ((bezugsgrenzeKw != null && bezugsgrenzeKw.compareTo(anschlussKva) > 0)
                || (einspeisegrenzeKw != null && einspeisegrenzeKw.compareTo(anschlussKva) > 0))) {
            return FEHLER_UEBER_ANSCHLUSS;
        }
        return null;
    }
}

File: services/optimization/voltpilot_optimization/grenze_aufloesung.py
"""Grenzen am Netzanschluss - der Python-Zwilling (UEMS AP-15 IP-3, Kasten W1).

Die EINE Lese-Regel: je Richtung gilt der ENGERE Wert aus Anlage und
Grenzblatt des Netzanschlusses; ohne gebundenen Netzanschluss oder ohne
gueltige Fassung der alte Wert der Anlage - dasselbe Objekt, damit eine Anlage
ohne Eintrag Byte fuer Byte bleibt.

Der Java-Zwilling ist ``services/api/.../uems/GrenzeAufloesung.java``; beide
fahren ``docs/contracts/v2/netzanschluss-grenze-vectors.json``. Wer die Regel
aendert, aendert die Vektor-Datei UND beide Zwillinge. Rein: ohne DB, ohne Uhr.
Tarif und Verguetung ziehen NICHT um (W1).
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date

QUELLE_ANLAGE = "anlage"
QUELLE_NETZANSCHLUSS = "netzanschluss"

FEHLER_UEBER_VEREINBART = "grenze_ueber_vereinbart"
FEHLER_UEBER_ANSCHLUSS = "grenze_ueber_anschluss"


@dataclass(frozen=True)
class Fassung:
    """Eine Fassung des Grenzblatts: gilt ab ihrem Tag bis zum Vortag der naechsten."""

    gueltig_ab: date
    einspeisegrenze_kw: float | None
    bezugsgrenze_kw: float | None
    einspeisegrenze_keine: bool = False


@dataclass(frozen=True)
class Wirksam:
    einspeisung_kw: float | None
    quelle_einspeisung: str | None
    bezug_kw: float | None
    quelle_bezug: str | None
    einspeisung_keine: bool = False


def fassung_am(fassungen, tag: date) -> Fassung | None:
    """Die Fassung mit dem spaetesten ersten Tag <= ``tag``; sonst ``None``."""
    gueltig = [f for f in (fassungen or ()) if f.gueltig_ab <= tag]
    return max(gueltig, key=lambda f: f.gueltig_ab) if gueltig else None


def engerer(anlage, netzanschluss):
    """Der engere Wert; gleich = der Wert der Anlage (dasselbe Objekt); einer fehlt = der andere."""
    if netzanschluss is None:
        return anlage
    if anlage is None:
        return netzanschluss
    return netzanschluss if netzanschluss < anlage else anlage


def _quelle(anlage, netzanschluss) -> str | None:
    wert = engerer(anlage, netzanschluss)
    if wert is None:
        return None
    return QUELLE_ANLAGE if wert is anlage else QUELLE_NETZANSCHLUSS


def aufloesen(
    einspeisung_kw, bezug_kw, gebunden: bool, fassungen, tag: date
) -> Wirksam:
    """Die wirksamen Grenzen am ``tag`` (Werte der Anlage, Bindung am Tag, Fassungen DIESES Anschlusses)."""
    f = fassung_am(fassungen, tag) if gebunden else None
    na_einspeisung = f.einspeisegrenze_kw if f is not None else None
    na_bezug = f.bezugsgrenze_kw if f is not None else None
    keine = einspeisung_kw is None and f is not None and f.einspeisegrenze_keine
    return Wirksam(
        einspeisung_kw=engerer(einspeisung_kw, na_einspeisung),
        quelle_einspeisung=QUELLE_NETZANSCHLUSS if keine else _quelle(einspeisung_kw, na_einspeisung),
        bezug_kw=engerer(bezug_kw, na_bezug),
        quelle_bezug=_quelle(bezug_kw, na_bezug),
        einspeisung_keine=keine,
    )


def plausibel(
    einspeisegrenze_kw, bezugsgrenze_kw, vereinbart_kw, anschluss_kva
) -> str | None:
    """Bezugsgrenze <= vereinbarte Leistung, jede Grenze <= Anschlussleistung; zuerst die vereinbarte."""
    if (
        bezugsgrenze_kw is not None
        and vereinbart_kw is not None
        and bezugsgrenze_kw > vereinbart_kw
    ):
        return FEHLER_UEBER_VEREINBART
    if anschluss_kva is not None and (
        (bezugsgrenze_kw is not None and bezugsgrenze_kw > anschluss_kva)
        or (einspeisegrenze_kw is not None and einspeisegrenze_kw > anschluss_kva)
    ):
        return FEHLER_UEBER_ANSCHLUSS
    return None
